core/src/main/scala/kanaloa/reactive/dispatcher/queue/QueueProcessor.scala

```diff
@@ -13,158 +10,151 @@ import kanaloa.reactive.dispatcher.{Backend, ResultChecker}
+
+    //if workers drop below minimum, we have a problem.
+
+    //if the Queue terminated, time to shut stuff down.
+    case Terminated(`queue`) ⇒
+      log.debug(s"Queue ${queue.path} is terminated")
+      if (workerPool.isEmpty) {
+        context stop self
+      } else {
+        workerPool.foreach(_ ! Worker.Retire)
+        delayedMsg(3.minutes, ShutdownTimeout) //TODO: hardcoded
+        context become shuttingDown(None)
+      }
+
+    case qs: QueryStatus ⇒ qs reply RunningStatus(workerPool)
+
+    //queue processor initiates shutdown of everyone.
+    case Shutdown(reportTo, timeout) ⇒
+      log.info("Commanded to shutdown. Shutting down")
+      queue ! Retire(timeout)
+      delayedMsg(timeout, ShutdownTimeout)
+      context become shuttingDown(reportTo)
   }
 
-  def shuttingDown(pool: WorkerPool, reportTo: Option[ActorRef]): Receive = {
+  def shuttingDown(reportTo: Option[ActorRef]): Receive = {
 
-    case Terminated(worker) if pool.contains(worker) ⇒
-      removeWorker(pool, worker, shuttingDown(_, reportTo), "successfully after command", reportTo)
+    case Terminated(worker) if workerPool.contains(worker) ⇒
+      removeWorker(worker)
+      if (workerPool.isEmpty) {
+        log.info(s"All Workers have terminated, QueueProcessor is shutting down")
+        reportTo.foreach(_ ! ShutdownSuccessfully)
+        context stop self
+      }
 
     case Terminated(_)   ⇒ //ignore other termination
 
     case qs: QueryStatus ⇒ qs reply ShuttingDown
 
     case ShutdownTimeout ⇒
       log.warning("Shutdown timed out, forcefully shutting down")
-      pool.foreach(_ ! PoisonPill)
+      workerPool.foreach(_ ! PoisonPill)
       context stop self
 
     case _ ⇒ //Ignore
   }
 
-  private def createWorker(index: Int): WorkerRef = {
-    val timestamp = LocalDateTime.now.toInstant(ZoneOffset.UTC).toEpochMilli
-    val worker = context.actorOf(
-      workerProp(queue),
-      s"worker-${queue.path.name}-$index-${timestamp}"
-    )
-    context watch worker
-    worker
+  def removeWorker(worker: ActorRef): Unit = {
+    context.unwatch(worker)
+    workerPool = workerPool - worker
   }
 
-  private def removeWorker(
-    pool:              WorkerPool,
-    worker:            WorkerRef,
-    nextContext:       WorkerPool ⇒ Receive,
-    finishWithMessage: String               = "",
-    reportToOnFinish:  Option[ActorRef]     = None
-  ): Unit = {
-    context unwatch worker
-    val newPool = pool - worker
-    if (!finishIfPoolIsEmpty(newPool, finishWithMessage, reportToOnFinish))
-      context become nextContext(newPool)
-
-  }
-
-  private def finishIfPoolIsEmpty(
-    pool:        WorkerPool,
-    withMessage: String,
-    reportTo:    Option[ActorRef] = None
-  ): Boolean = {
-    val finishes = pool.isEmpty
-    if (finishes) {
-      log.info(s"Queue Processor is shutdown $withMessage")
-      reportTo.foreach(_ ! ShutdownSuccessfully)
-      context stop self
+  private def createRoutee(): Unit = {
```

Comment:
it might be better to call it retrieveRoutee, this might gives the false impression that it creates a new Routee every time called.